optuna/batch/trial.py

```diff
@@ -0,0 +1,154 @@
+import abc
+import datetime
+from typing import Any
+from typing import Dict
+from typing import Optional
+from typing import Sequence
+from typing import Union
+import warnings
+
+import numpy as np
+
+import optuna
+
+
+class BaseBatchTrial(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def _get_trials(
+        self,
+    ) -> Union[
+        Sequence["optuna.trial.Trial"],
+        Sequence["optuna.multi_objective.trial.MultiObjectiveTrial"],
+    ]:
+        raise NotImplementedError
+
+    def suggest_float(
+        self,
+        name: str,
+        low: float,
+        high: float,
+        *,
+        step: Optional[float] = None,
+        log: bool = False
+    ) -> np.ndarray:
+        return np.array(
+            [t.suggest_float(name, low, high, step=step, log=log) for t in self._get_trials()]
+        )
+
+    def suggest_uniform(self, name: str, low: float, high: float) -> np.ndarray:
+        return np.array([t.suggest_uniform(name, low, high) for t in self._get_trials()])
```

Comment:
In my understanding, one of the motivations behind batched sampling is performance gains under vectorized computations (with e.g. `numpy`). The objective function will be able to utilize this. However, samplers will not. Maybe it's something further down the roadmap but I just thought that a "batched sampler" would ideally be vectorized as well, allocating some continuous memory for the entire batch in a single allocation instead of using batch size number of allocations and a concatenation.

Reply:
Thank you for your suggestion, but I think the `batched sample` is not necessary to implement the vectorized computation of parameter suggestions according to the comment of https://github.com/optuna/optuna/pull/1729#discussion_r488433123. The current design of samplers enables users to suggest trials in a batched manner if they add a parameter queue in the samplers.